Platform refactoring and macOS implementation

s1gtrap opened the pull request:
Splits the codebase up into platform-specific modules, along with a few minor API changes (like not exporting Windows dependent `POINT` type). Should help a bit in implementing it for other Unixy platforms as well, though keep ind mind that `macos` is part of the `unix` family, so should probably have some `not(target_os = "macos")` magic in there when time comes.

Using it on macOS is a little weird, as it needs the Accessibility entitlement. After a bit of pondering why it wasn't working, I added iTerm to the Accessibility permission under Security & Privacy settings and it worked wonders.

Tried and tested on Windows 10 and macOS Catalina.

s1gtrap committed:
macOS impl

Splits the codebase up in platforms, each residing in their respective
sys modules, and makes dependencies platform dependent.

s1gtrap committed:
bump to 0.3

Bumps the minor, due to API changes.

DankDumpster commented:
Wow, this is so much work. I'm really happy that you put in all this time. I'll look through all of it in about 1 or 2 hours. Thanks again!

AltF02 commented on src/sys/macos.rs:
```diff
@@ -0,0 +1,133 @@
+            _ if button == LEFT => Ok((CGEventType::LeftMouseDown, CGMouseButton::Left)),
+            _ if button == MIDDLE => Ok((CGEventType::LeftMouseDown, CGMouseButton::Center)),
+            _ if button == RIGHT => Ok((CGEventType::RightMouseDown, CGMouseButton::Right)),
+            _ => Err(Box::new(Error::InvalidButtonStr(button))),
```
I'm not that experienced in rust yet but may I ask why this is better than doing?

```rust

        let (event_type, mouse_button) = match button {
            LEFT => Ok((CGEventType::LeftMouseDown, CGMouseButton::Left)),
            MIDDLE => Ok((CGEventType::LeftMouseDown, CGMouseButton::Center)),
            RIGHT => Ok((CGEventType::RightMouseDown, CGMouseButton::Right)),
            _ => Err(Box::new(Error::InvalidButtonStr(button))),
        }

s1gtrap replied:
Heh, the problem with your snippet is that `LEFT`, despite it being a predefined `const`, doesn't restrict the pattern like in OCaml. Instead the string slice being matched  is bound to a new variable called `LEFT`, regardless of its actual contents. In other words, the latter three patterns are unreachable while `"left"`, `"right"`, and even `"🍆"` matches to a left button press.

It would be possible to do something like this instead,

```
let (event_type, mouse_button) = match button {
    "left" => Ok((CGEventType::LeftMouseDown, CGMouseButton::Left)),
    "middle" => Ok((CGEventType::LeftMouseDown, CGMouseButton::Center)),
    "right" => Ok((CGEventType::RightMouseDown, CGMouseButton::Right)),
    _ => Err(Box::new(Error::InvalidButtonStr(button))),
}
```

but that kind of defeats the purpose of declaring the `const`s in the first place.

I think it would be better to declare an enum instead, so something like,
```
enum Keys {
    Left,
    Right,
    // ...
}
```
and match over that, instead of taking a `&str`. That would also mean no panics, and in my opinion cleaner code. 😉

But I'll leave that up to you. I thought about including it in this PR or creating another, but it felt like a bigger overhaul than needed.

AltF02 replied:
Thank you for all the info 😄

AltF02 commented on src/sys/macos.rs:
```diff
@@ -0,0 +1,133 @@
+            _ if button == RIGHT => Ok((CGEventType::RightMouseUp, CGMouseButton::Right)),
+            _ => Err(Box::new(Error::InvalidButtonStr(button))),
+        }?;
+
```
Is it possible to combine the duplicate from the latter comment into one method?

s1gtrap replied:
I actually thought about that, but given that one is for up and the other for down variants the only thing I can imagine is more state (like passing a `bool` along to signify up/down) or some nasty `concat!` business. But hey, you got me to look up `CKMouseButton`, and turns out that it has a different `CGEventType` than the other two!

AltF02 reviewed (changes requested):
Looks great if possible could you apply these changes? Thanks!

s1gtrap committed:
CGMouseButton.center is otherMouseUp/Down event

kCGMouseButtonCenter being defined as 2 actually makes it a
kCGEventOtherMouseUp/Down event given than it spans button 2-31.

AltF02 reviewed (approved):
Lgtm, again thank you for all this work
